l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,678 @@
+    mapping(uint256 chainId => mapping(address l1Token => uint256 balance)) internal chainBalance;
+
+    /// @dev Indicates whether the hyperbridging is enabled for a given chain.
+    mapping(uint256 chainId => bool enabled) internal hyperbridgingEnabled;
+
+    /// @notice Checks that the message sender is the bridgehub
+    modifier onlyBridgehub() {
+        require(msg.sender == address(bridgehub), "ShB not BH");
+        _;
+    }
+
+    /// @notice Checks that the message sender is the bridgehub or Era
+    modifier onlyBridgehubOrEra(uint256 _chainId) {
+        require(
+            (msg.sender == address(bridgehub)) ||
+                ((_chainId == ERA_CHAIN_ID) && (msg.sender == bridgehub.getStateTransition(_chainId))),
+            "L1SharedBridge: not bridgehub or era chain"
+        );
+        _;
+    }
+
+    /// @notice Checks that the message sender is the legacy bridge
+    modifier onlyLegacyBridge() {
+        require(msg.sender == address(legacyBridge), "ShB not legacy bridge");
+        _;
+    }
+
+    /// @dev Contract is expected to be used as proxy implementation.
+    /// @dev Initialize the implementation to prevent Parity hack.
+    constructor(
+        address payable _l1WethAddress,
+        IBridgehub _bridgehub,
+        IL1ERC20Bridge _legacyBridge
+    ) reentrancyGuardInitializer {
+        l1WethAddress = _l1WethAddress;
+        bridgehub = _bridgehub;
+        legacyBridge = _legacyBridge;
+    }
+
+    /// @dev Initializes a contract bridge for later use. Expected to be used in the proxy
+    /// @param _owner Address which can change L2 token implementation and upgrade the bridge
+    /// implementation. The owner is the Governor and separate from the ProxyAdmin from now on, so that the Governor can call the bridge
+    function initialize(
+        address _owner,
+        uint256 _eraIsEthWithdrawalFinalizedStorageSwitchBatchNumber
+    ) external reentrancyGuardInitializer reinitializer(2) {
+        _transferOwnership(_owner);
+        require(_owner != address(0), "ShB owner 0");
+
+        eraIsEthWithdrawalFinalizedStorageSwitchBatchNumber = _eraIsEthWithdrawalFinalizedStorageSwitchBatchNumber;
+
+        l2BridgeAddress[ERA_CHAIN_ID] = ERA_ERC20_BRIDGE_ADDRESS;
+    }
+
+    /// @dev Initializes the l2Bridge address by governance for a specific chain
+    function initializeChainGovernance(uint256 _chainId, address _l2BridgeAddress) external onlyOwner {
+        l2BridgeAddress[_chainId] = _l2BridgeAddress;
+    }
+
+    /// @notice used by bridgehub to aquire mintValue. If l2Tx fails refunds are sent to refundrecipient on L2
+    /// we also use it to keep to track each chain's assets
+    function bridgehubDepositBaseToken(
+        uint256 _chainId,
+        address _prevMsgSender,
+        address _l1Token,
+        uint256 _amount
+    ) external payable onlyBridgehubOrEra(_chainId) {
+        require(_amount != 0, "4T"); // empty deposit amount
```

Comment:
I agree with most (maybe all) of the other comments and am fixing them. Here however I don't think we should allow 0 amount, this means the message value and l2GasFees would be zero. This is already checked in the Mailbox to not be allowed.

Reply:
Here: 
https://github.com/matter-labs/era-contracts/blob/6f322ceda56cf0a211f0ab3bab83f1fefa688d64/l1-contracts/contracts/zksync/facets/Mailbox.sol#L251
Unless the tx is free, which is only for system transactions.

Reply:
True, but I just think this restriction is not fundamental as it cames from Era gas model for L1 -> L2 transactions
@kelemeno Do you think hyperchains may allow free L1->L2 transactions? Now or in the future

Reply:
This is like bikeshedding, so can discuss later

Reply:
Ah ok, true it is not fundamental. I agree.

Reply:
Sorry to jump into this discussion

Had a discussion with @kelemeno yesterday and to allow `baseToken` deposit amount to be 0, we need to modify the `bootloader` to support fee payment during a L1 transaction processing by the transaction `from` account rather than deducting from the `valueToMint`

While the possibility to have the L1->L2 transaction be subsidised by the chain ( 0 baseToken deposit) is attractive for us, do you think it will be possible to update the bootloader contract once the chain is started?

Please let me know about the conclusion of the discussion later on @kelemeno